### PR TITLE
Add Windows Arm64 Nightly Builds

### PR DESCRIPTION
**Pull request by iremyux**
This PR creates 3 new worklflows for Windows Arm64 target. The workflows and outputs can be reviewed at the following links:
https://github.com/pytorch/pytorch/actions/workflows/generated-windows-arm64-binary-libtorch-release-nightly.yml
https://github.com/pytorch/pytorch/actions/workflows/generated-windows-arm64-binary-libtorch-debug-nightly.yml
https://github.com/pytorch/pytorch/actions/workflows/generated-windows-arm64-binary-wheel-nightly.yml

### CODE REVIEW

File: .github/workflows/_win-arm64-build.yml [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+      test-matrix: ${{ steps.filter.outputs.test-matrix }}
+    defaults:
+      run:
+        shell: bash
+    steps:
+      - name: configure aws credentials
+        id: aws_creds
+        uses: aws-actions/configure-aws-credentials@v4
+        with:
+          role-to-assume: arn:aws:iam::308535385114:role/gha_workflow_s3_and_ecr_read_only
```
Comment (huydhn) [inline .github/workflows/_win-arm64-build.yml:91]: Could you try to use `gha_workflow_sscache` instead?
```suggestion
          role-to-assume: arn:aws:iam::308535385114:role/gha_workflow_sscache
```

File: .ci/pytorch/win-arm64-test-helpers/build_pytorch.bat [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+sccache --zero-stats
+
+python setup.py bdist_wheel
+if errorlevel 1 goto fail
+if not errorlevel 0 goto fail
+sccache --show-stats
+python -c "import os, glob; os.system('python -mpip install --no-index --no-deps ' + glob.glob('dist/*.whl')[0])"
+(
+  if "%BUILD_ENVIRONMENT%"=="" (
+    echo NOTE: To run `import torch`, please make sure to activate the conda environment by running `call %CONDA_PARENT_DIR%\Miniconda3\Scripts\activate.bat %CONDA_PARENT_DIR%\Miniconda3` in Command Prompt before running Git Bash.
```
Comment (atalman) [inline .ci/pytorch/win-arm64-test-helpers/build_pytorch.bat:73]: Hi @iremyux do we really need to install conda here ? We are trying to get rid of conda dependency

File: .ci/pytorch/windows/arm64/smoke_test.bat [hunk 3]
```diff
@@ -34,15 +27,28 @@ echo Checking that basic CNN works...
 python %PYTORCH_ROOT%\.ci\pytorch\test_example_code\cnn_smoke_win_arm64.py
 if errorlevel 1 exit /b 1
 
+push %PYTORCH_ROOT%\test
+
+set CORE_TEST_LIST=test_autograd.py test_nn.py test_torch.py
```
Comment (atalman) [inline .ci/pytorch/windows/arm64/smoke_test.bat:32]: Can we also call 
```https://github.com/pytorch/pytorch/blob/main/.ci/pytorch/smoke_test/smoke_test.py```

File: .ci/pytorch/win-arm64-test-helpers/build_pytorch.bat (hunk 2 above)
Comment (iremyux) [reply .ci/pytorch/win-arm64-test-helpers/build_pytorch.bat:73]: hey, yeap i think we missed it, conda is not available for arm64 anyway, I removed the installation.

File: .ci/pytorch/windows/arm64/smoke_test.bat (hunk 3 above)
Comment (iremyux) [reply .ci/pytorch/windows/arm64/smoke_test.bat:30]: done. results can be seen here:

**Comment by iremyux** [comment]
Removing ciflow/binaries label since windows-arm64-binary-wheel, windows-arm64-binary-libtorch-debug and windows-arm64-binary-libtorch-release are all stable and green now. No need to trigger them anymore.

**Comment by iremyux** [comment]
Removing ciflow/binaries label since nightlies are all stable and green now. Here are the last runs: 
[windows-arm64-binary-wheel · pytorch/pytorch@65d08a2](https://github.com/pytorch/pytorch/actions/runs/13542265024)
[windows-arm64-binary-libtorch-release · pytorch/pytorch@415728c](https://github.com/pytorch/pytorch/actions/runs/13519381877)
[windows-arm64-binary-libtorch-debug · pytorch/pytorch@415728c](https://github.com/pytorch/pytorch/actions/runs/13519381887)

**Comment by iremyux** [comment]
last successful arm64 nightly workflows: 
[windows-arm64-binary-libtorch-release · pytorch/pytorch@415728c](https://github.com/pytorch/pytorch/actions/runs/13519381877)
[windows-arm64-binary-libtorch-debug · pytorch/pytorch@415728c](https://github.com/pytorch/pytorch/actions/runs/13519381887)
[windows-arm64-binary-wheel · pytorch/pytorch@97dafd8](https://github.com/pytorch/pytorch/actions/runs/13589031393)

File: .ci/pytorch/test_example_code/rnn_smoke_win_arm64.py [hunk 4]
```diff
@@ -6,8 +6,7 @@
 import torch
 import torch.nn as nn
 
-
 rnn = nn.RNN(10, 20, 2)
 inputs = torch.randn(5, 3, 10)
 h0 = torch.randn(2, 3, 20)
-output, hn = rnn(inputs, h0)
+output, hn = rnn(inputs, h0)
```
Comment (Blackhex) [inline .ci/pytorch/test_example_code/rnn_smoke_win_arm64.py:12]: This change and the change in `.ci/pytorch/test_example_code/cnn_smoke_win_arm64.py` file seem to be unnecessary and not relevant to the rest of the PR.

File: .ci/pytorch/windows/arm64/bootstrap_buildtools.bat [hunk 5]
```diff
@@ -1,4 +1,5 @@
 @echo off
+setlocal enabledelayedexpansion
```
Comment (Blackhex) [inline .ci/pytorch/windows/arm64/bootstrap_buildtools.bat:2]: What was the issue with the original version of the script so this line had to be added?

File: .ci/pytorch/windows/arm64/bootstrap_buildtools.bat [hunk 6]
```diff
@@ -18,31 +19,46 @@ curl -L -o "%INSTALLER_FILE%" %DOWNLOAD_URL%
 echo Installing Visual Studio Build Tools with C++ components...
 echo Installing MSVC %MSVC_VERSION%
 if "%MSVC_VERSION%" == "latest" (
-    "%INSTALLER_FILE%" --norestart --nocache --quiet --wait --installPath "%DEPENDENCIES_DIR%\VSBuildTools" ^
+    "%INSTALLER_FILE%" --norestart --quiet --wait --installPath "%DEPENDENCIES_DIR%\VSBuildTools" ^
```
Comment (Blackhex) [inline .ci/pytorch/windows/arm64/bootstrap_buildtools.bat:22]: When such significant change of the list is done, I 'd sort them by component names alphabetically as well to make the future maintenance easier.

File: .ci/pytorch/windows/arm64/bootstrap_buildtools.bat [hunk 7]
```diff
@@ -18,31 +19,46 @@ curl -L -o "%INSTALLER_FILE%" %DOWNLOAD_URL%
...
         --add Microsoft.VisualStudio.Component.VC.CMake.Project ^
         --add Microsoft.VisualStudio.Component.VC.14.36.17.6.ARM64 ^
         --add Microsoft.VisualStudio.Component.VC.14.36.17.6.x86.x64
 )
 
+echo exitcode = %errorlevel%
+
 :: Check if installation was successful
 if %errorlevel% neq 0 (
-    echo "Failed to install Visual Studio Build Tools with C++ components. (exitcode = %errorlevel%)"
```
Comment (Blackhex) [inline .ci/pytorch/windows/arm64/bootstrap_buildtools.bat:45]: This and the change of the previous lines does not make much sense: When the `%errorlevel%` is non-zero it's better to print the value together with the error message, when it's zero it's kind of unecessary to print it.

File: .ci/pytorch/windows/arm64/smoke_test.bat [hunk 8]
```diff
@@ -60,6 +60,4 @@ if ERRORLEVEL 1 exit /b 1
 .\simple-torch-test.exe
 if ERRORLEVEL 1 exit /b 1
 
-:end
-set "PATH=%ORIG_PATH%"
```
Comment (Blackhex) [inline .ci/pytorch/windows/arm64/smoke_test.bat:64]: This line was there for a reason. Is that reason no longer valid? If yes, saving the path to `ORIG_PATH` at the beginning fo the script should be removed as well.

**Review by Blackhex — COMMENTED** [review]
Overall looks good, I left some minor comments.

File: .ci/pytorch/test_example_code/rnn_smoke_win_arm64.py (hunk 4 above)
Comment (iremyux) [reply .ci/pytorch/test_example_code/rnn_smoke_win_arm64.py]: you're right, they do not appear as changed now.

File: .ci/pytorch/windows/arm64/bootstrap_buildtools.bat (hunk 7 above)
Comment (Blackhex) [reply .ci/pytorch/windows/arm64/bootstrap_buildtools.bat:45]: OK, thank you for the explanation.

File: .ci/pytorch/windows/arm64/bootstrap_buildtools.bat (hunk 6 above)
Comment (malfet) [reply .ci/pytorch/windows/arm64/bootstrap_buildtools.bat:21]: @ozanMSFT does it mean that this is still a draft PR or a final version?

File: .ci/pytorch/windows/arm64/bootstrap_buildtools.bat [hunk 9]
```diff
@@ -18,31 +18,49 @@ curl -L -o "%INSTALLER_FILE%" %DOWNLOAD_URL%
...
         --add Microsoft.VisualStudio.Component.Windows11SDK.22621 ^
         --add Microsoft.VisualStudio.Component.VC.ASAN ^
         --add Microsoft.VisualStudio.Component.VC.CMake.Project ^
+        --add Microsoft.VisualStudio.Component.VC.CoreBuildTools ^
+        --add Microsoft.VisualStudio.Component.VC.CoreIde ^
+        --add Microsoft.VisualStudio.Component.VC.Redist.14.Latest ^
+        --add Microsoft.VisualStudio.Component.VC.Tools.ARM64EC ^
         --add Microsoft.VisualStudio.Component.VC.Tools.ARM64 ^
         --add Microsoft.VisualStudio.Component.VC.Tools.x86.x64
 ) else if "%MSVC_VERSION%" == "14.40" (
```
Comment (malfet) [inline .ci/pytorch/windows/arm64/bootstrap_buildtools.bat:33]: Why do you need to support so many different versions of the toolchain right away? For CI you can always pick one, can't you?

File: .ci/pytorch/windows/arm64/smoke_test.bat [hunk 10]
```diff
@@ -34,15 +20,29 @@ echo Checking that basic CNN works...
...
+python %PYTORCH_ROOT%\.ci\pytorch\smoke_test\smoke_test.py --package torchonly
+
+echo Running test_autograd.oy, test_nn.py, test_torch.py...
+cd %PYTORCH_ROOT%\test
+set CORE_TEST_LIST=test_autograd.py test_nn.py test_torch.py
+
+for %%t in (%CORE_TEST_LIST%) do (
+    echo Running test: %%t
+    python %%t --verbose --save-xml --use-pytest -vvvv -rfEsxXP -p no:xdist
+)
```
Comment (malfet) [inline .ci/pytorch/windows/arm64/smoke_test.bat:33]: We usually don't do full test runs during nightly builds, this should be part of CI pipeline

**Review by malfet — CHANGES_REQUESTED** [review]
Please do not add testing to CD builds, as tests should have been run during CI.
Otherwise, looks OK to me, though I don't understand why it's necessary to keep adding new batch scripts in 2025. Aren't PowerShell available by default on all Windows-on-ARM systems?

File: .ci/pytorch/windows/arm64/bootstrap_buildtools.bat (hunk 9 above)
Comment (malfet) [reply .ci/pytorch/windows/arm64/bootstrap_buildtools.bat:33]: Since we are not testing those, do you mind deleting dead code from the script?
Also, I would advice against latest, as this would results in CD builds starting to fail on one day and it makes it really hard to figure out that it was due to the version change

File: .github/scripts/generate_ci_workflows.py [hunk 11]
```diff
@@ -261,6 +262,85 @@ class OperatingSystem:
...
 ]
 
+WINDOWS_ARM64_BINARY_BUILD_WORKFLOWS = [
+    BinaryBuildWorkflow(
+        os=OperatingSystem.WINDOWS_ARM64,
+        package_type="wheel",
+        build_configs=generate_binary_build_matrix.generate_wheels_matrix(
+            OperatingSystem.WINDOWS_ARM64,
+            arches=["cpu"],
+            python_versions=["3.12"],
```
Comment (malfet) [inline .github/scripts/generate_ci_workflows.py:272]: I guess windows-on-arm users only care about Python-3.12.. Oh well.

File: .github/templates/windows_arm64_binary_build_workflow.yml.j2 [hunk 12]
```diff
@@ -0,0 +1,208 @@
+{% import 'common.yml.j2' as common %}
```
Comment (malfet) [inline .github/templates/windows_arm64_binary_build_workflow.yml.j2:1]: This whole copy-n-paste makes me sad, but whatever

**Review by malfet — APPROVED** [review]
I'm sad about amount of copy-and-paste (and batch scripts), but sure, let's land it as is to get some signal.

But let's plan to revert this change(in say a month from now) if no dedup/batch->powershell migration happens

**Review by malfet — CHANGES_REQUESTED** [review]
Hmm, I didn't notice you've also added `main` builds.
Please remove them, no need to build Windows-on-ARM binaries for every trunk commit

**Review by malfet — APPROVED** [review]

File: .github/templates/windows_arm64_binary_build_workflow.yml.j2 [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+      !{{ set_runner_specific_vars() }}
+      - name: Bootstrap folders
+        shell: cmd
+        run: |
+          mkdir "%NIGHTLIES_PYTORCH_ROOT%"
+          mkdir "%PYTORCH_FINAL_PACKAGE_DIR%"
+      - name: Git checkout PyTorch
+        uses: actions/checkout@v4
+        with:
+          repository:  "iremyux/pytorch"
```
Comment (seemethere) [inline .github/templates/windows_arm64_binary_build_workflow.yml.j2:88]: This will probably need to change

File: .github/pytorch-probot.yml [hunk 14]
```diff
@@ -23,6 +23,7 @@ ciflow_push_tags:
 - ciflow/xpu
 - ciflow/torchbench
 - ciflow/autoformat
+- ciflow/win-arm64
```
Comment (malfet) [inline .github/pytorch-probot.yml:26]: Why are you adding this one? `ciflow/` labels are usually reserved for CI, for CD everything should be triggered by `ciflow/binaries`

**Review by malfet — CHANGES_REQUESTED** [review]
Please do not trigger CD, on `ciflow/win-arm64` labels (`ciflow/binary_wheels` should be sufficient)

File: .github/scripts/generate_ci_workflows.py (hunk 11 above)
Comment (iremyux) [reply .github/scripts/generate_ci_workflows.py:272]: I think we discussed this on yesterday's meeting.

File: .github/templates/windows_arm64_binary_build_workflow.yml.j2 (hunk 12 above)
Comment (iremyux) [reply .github/templates/windows_arm64_binary_build_workflow.yml.j2:1]: I think we discussed this on yesterday's meeting.

File: .github/templates/windows_arm64_binary_build_workflow.yml.j2 (hunk 13 above)
Comment (iremyux) [reply .github/templates/windows_arm64_binary_build_workflow.yml.j2]: Done

**Comment by iremyux** [comment]
@malfet @seemethere 
All of the reviews are addressed and resolved, can you please take a look again?

**Review by malfet — APPROVED** [review]
LGTM, but I'm filing a followup task to https://github.com/pytorch/pytorch/issues/148776 and propose to revert this PR if this is not closed within next say 30 days
